File: features/user/user_identification/src/main/java/com/tokopedia/useridentification/view/fragment/UserIdentificationFormFinalFragment.java
package com.tokopedia.useridentification.view.fragment;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.tokopedia.abstraction.base.app.BaseMainApplication;
import com.tokopedia.abstraction.base.view.activity.BaseStepperActivity;
import com.tokopedia.abstraction.base.view.fragment.BaseDaggerFragment;
import com.tokopedia.abstraction.base.view.listener.StepperListener;
import com.tokopedia.abstraction.common.utils.image.ImageHandler;
import com.tokopedia.applink.RouteManager;
import com.tokopedia.user_identification_common.KycCommonUrl;
import com.tokopedia.useridentification.R;
import com.tokopedia.useridentification.analytics.UserIdentificationAnalytics;
import com.tokopedia.useridentification.di.DaggerUserIdentificationComponent;
import com.tokopedia.useridentification.di.UserIdentificationComponent;
import com.tokopedia.useridentification.view.activity.UserIdentificationCameraActivity;
import com.tokopedia.useridentification.view.activity.UserIdentificationFormActivity;
import com.tokopedia.useridentification.view.activity.UserIdentificationInfoActivity;
import com.tokopedia.useridentification.view.listener.UserIdentificationUploadImage;
import com.tokopedia.useridentification.view.viewmodel.UserIdentificationStepperModel;

import java.io.File;

import javax.inject.Inject;

import static com.tokopedia.user_identification_common.KYCConstant.EXTRA_STRING_IMAGE_RESULT;
import static com.tokopedia.user_identification_common.KYCConstant.REQUEST_CODE_CAMERA_FACE;
import static com.tokopedia.user_identification_common.KYCConstant.REQUEST_CODE_CAMERA_KTP;
import static com.tokopedia.useridentification.view.fragment.UserIdentificationCameraFragment.PARAM_VIEW_MODE_FACE;
import static com.tokopedia.useridentification.view.fragment.UserIdentificationCameraFragment.PARAM_VIEW_MODE_KTP;

/**
 * @author by alvinatin on 15/11/18.
 */

public class UserIdentificationFormFinalFragment extends BaseDaggerFragment
        implements UserIdentificationUploadImage.View, UserIdentificationFormActivity.Listener {

    public static final String PARAM_PROJECTID = "TRADEIN_PROJECT";

    private ImageView imageKtp;
    private ImageView imageFace;
    private TextView buttonKtp;
    private TextView buttonFace;
    private TextView info;
    private TextView uploadButton;
    private View progressBar;
    private UserIdentificationStepperModel stepperModel;

    private StepperListener stepperListener;
    private UserIdentificationAnalytics analytics;

    private int projectId;

    @Inject
    UserIdentificationUploadImage.Presenter presenter;

    public static Fragment createInstance(int projectid) {
        UserIdentificationFormFinalFragment fragment = new UserIdentificationFormFinalFragment();
        Bundle bundle = new Bundle();
        bundle.putInt(PARAM_PROJECTID, projectid);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getContext() instanceof StepperListener) {
            stepperListener = (StepperListener) getContext();
        }
        if (getArguments() != null && savedInstanceState == null) {
            stepperModel = getArguments().getParcelable(BaseStepperActivity.STEPPER_MODEL_EXTRA);
        } else if (savedInstanceState != null) {
            stepperModel = savedInstanceState.getParcelable(BaseUserIdentificationStepperFragment
                    .EXTRA_KYC_STEPPER_MODEL);
        }
        if (getActivity() != null) {
            analytics = UserIdentificationAnalytics.createInstance(getActivity()
                    .getApplicationContext());
        }
        projectId = getArguments().getInt(PARAM_PROJECTID, -1);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putParcelable(BaseUserIdentificationStepperFragment.EXTRA_KYC_STEPPER_MODEL, stepperModel);
        super.onSaveInstanceState(outState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_user_identification_final, container, false);
        if (projectId == 4) //TradeIn project Id
            uploadButton.setText(R.string.upload_button_tradein);
        initView(view);
        setContentView();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        hideLoading();
        analytics.eventViewFinalForm();
    }

    @Override
    protected void initInjector() {
        if (getActivity() != null) {
            UserIdentificationComponent daggerUserIdentificationComponent =
                    DaggerUserIdentificationComponent.builder()
                            .baseAppComponent(((BaseMainApplication) getActivity().getApplication()).getBaseAppComponent())
                            .build();

            daggerUserIdentificationComponent.inject(this);
            presenter.attachView(this);
        }
    }

    private void setContentView() {
        setImageKtp(stepperModel.getKtpFile());
        setImageFace(stepperModel.getFaceFile());
        generateLink();
        buttonKtp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                analytics.eventClickChangeKtpFinalFormPage();
                Intent intent = UserIdentificationCameraActivity.createIntent(getContext(),
                        PARAM_VIEW_MODE_KTP);
                startActivityForResult(intent, REQUEST_CODE_CAMERA_KTP);
            }
        });

        buttonFace.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                analytics.eventClickChangeSelfieFinalFormPage();
                Intent intent = UserIdentificationCameraActivity.createIntent(getContext(),
                        PARAM_VIEW_MODE_FACE);
                startActivityForResult(intent, REQUEST_CODE_CAMERA_FACE);
            }
        });

        uploadButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                analytics.eventClickUploadPhotos();
                uploadImage();
            }
        });
        if (getActivity() instanceof UserIdentificationFormActivity) {
            ((UserIdentificationFormActivity) getActivity())
                    .updateToolbarTitle(getString(R.string.title_kyc_form_upload));
        }
    }

    private void uploadImage() {
        showLoading();
        presenter.uploadImage(stepperModel, projectId);
    }

    private void setImageKtp(String imagePath) {
        File ktpFile = new File(imagePath);
        if (ktpFile.exists()) {
            ImageHandler.loadImageFromFile(getContext(), imageKtp, ktpFile);
        }
    }

    private void setImageFace(String imagePath) {
        File faceFile = new File(imagePath);
        if (faceFile.exists()) {
            ImageHandler.loadImageFromFile(getContext(), imageFace, faceFile);
        }
    }

    private void initView(View view) {
        imageKtp = view.findViewById(R.id.image_ktp);
        imageFace = view.findViewById(R.id.image_face);
        buttonKtp = view.findViewById(R.id.change_ktp);
        buttonFace = view.findViewById(R.id.change_face);
        info = view.findViewById(R.id.text_info);
        uploadButton = view.findViewById(R.id.upload_button);
        progressBar = view.findViewById(R.id.progress_bar);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK && data != null) {
            String imagePath = data.getStringExtra(EXTRA_STRING_IMAGE_RESULT);
            switch (requestCode) {
                case REQUEST_CODE_CAMERA_KTP:
                    stepperModel.setKtpFile(imagePath);
                    setImageKtp(imagePath);
                    break;
                case REQUEST_CODE_CAMERA_FACE:
                    stepperModel.setFaceFile(imagePath);
                    setImageFace(imagePath);
                    break;
                default:
                    break;
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected String getScreenName() {
        return null;
    }

    private void generateLink() {
        ClickableSpan clickableSpan = new ClickableSpan() {
            @Override
            public void onClick(View widget) {
                analytics.eventClickTermsFinalFormPage();
                RouteManager.route(getActivity(), KycCommonUrl.APPLINK_TERMS_AND_CONDITION);
            }

            @Override
            public void updateDrawState(TextPaint ds) {
                super.updateDrawState(ds);
                ds.setUnderlineText(false);
                ds.setColor(getResources().getColor(R.color.tkpd_main_green));
            }
        };

        SpannableString infoText = new SpannableString(info.getText());
        String linked = getResources().getString(R.string.terms_and_condition);
        int startIndex = info.getText().toString().indexOf(linked);
        infoText.setSpan(clickableSpan, startIndex, startIndex + linked.length(), Spanned
                .SPAN_EXCLUSIVE_EXCLUSIVE);
        info.setHighlightColor(Color.TRANSPARENT);
        info.setMovementMethod(LinkMovementMethod.getInstance());
        info.setText(infoText, TextView.BufferType.SPANNABLE);
    }

    @Override
    public void goToNextActivity() {
        Intent intent = new Intent(getContext(), UserIdentificationInfoActivity.class);
        startActivity(intent);
    }

    @Override
    public void onSuccessUpload() {
        hideLoading();
        getActivity().setResult(Activity.RESULT_OK);
        stepperListener.finishPage();
    }

    @Override
    public void onErrorUpload(String error) {
        hideLoading();
        if (getActivity() instanceof UserIdentificationFormActivity) {
            ((UserIdentificationFormActivity) getActivity()).showError(error, this::uploadImage);
        }
    }

    @Override
    public void showLoading() {
        imageKtp.setVisibility(View.GONE);
        imageFace.setVisibility(View.GONE);
        buttonKtp.setVisibility(View.GONE);
        buttonFace.setVisibility(View.GONE);
        info.setVisibility(View.GONE);
        uploadButton.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        imageKtp.setVisibility(View.VISIBLE);
        imageFace.setVisibility(View.VISIBLE);
        buttonKtp.setVisibility(View.VISIBLE);
        buttonFace.setVisibility(View.VISIBLE);
        info.setVisibility(View.VISIBLE);
        uploadButton.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public void trackOnBackPressed() {
        analytics.eventClickBackFinalForm();
    }
}

File: features/user/kyc_centralized/src/main/java/com/tokopedia/kyc_centralized/view/fragment/UserIdentificationFormFinalFragment.kt
package com.tokopedia.kyc_centralized.view.fragment

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.SpannableString
import android.text.Spanned
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.tokopedia.abstraction.base.app.BaseMainApplication
import com.tokopedia.abstraction.base.view.activity.BaseStepperActivity
import com.tokopedia.abstraction.base.view.fragment.BaseDaggerFragment
import com.tokopedia.abstraction.base.view.listener.StepperListener
import com.tokopedia.abstraction.common.utils.image.ImageHandler
import com.tokopedia.abstraction.common.utils.snackbar.NetworkErrorHelper
import com.tokopedia.applink.RouteManager
import com.tokopedia.applink.internal.ApplinkConstInternalGlobal
import com.tokopedia.kotlin.extensions.view.toEmptyStringIfNull
import com.tokopedia.kyc_centralized.R
import com.tokopedia.kyc_centralized.data.model.response.KycData
import com.tokopedia.kyc_centralized.di.DaggerUserIdentificationCommonComponent
import com.tokopedia.kyc_centralized.util.ImageEncryptionUtil
import com.tokopedia.kyc_centralized.util.KycUploadErrorCodeUtil
import com.tokopedia.kyc_centralized.view.activity.UserIdentificationCameraActivity.Companion.createIntent
import com.tokopedia.kyc_centralized.view.activity.UserIdentificationFormActivity
import com.tokopedia.kyc_centralized.view.listener.UserIdentificationUploadImage
import com.tokopedia.kyc_centralized.view.model.UserIdentificationStepperModel
import com.tokopedia.kyc_centralized.view.viewmodel.KycUploadViewModel
import com.tokopedia.logger.ServerLogger
import com.tokopedia.logger.utils.Priority
import com.tokopedia.remoteconfig.FirebaseRemoteConfigImpl
import com.tokopedia.remoteconfig.RemoteConfig
import com.tokopedia.remoteconfig.RemoteConfigKey
import com.tokopedia.unifycomponents.UnifyButton
import com.tokopedia.usecase.coroutines.Fail
import com.tokopedia.usecase.coroutines.Success
import com.tokopedia.user_identification_common.KYCConstant
import com.tokopedia.user_identification_common.KycCommonUrl
import com.tokopedia.user_identification_common.KycUrl
import com.tokopedia.user_identification_common.analytics.UserIdentificationCommonAnalytics
import kotlinx.android.synthetic.main.layout_kyc_upload_error.*
import com.tokopedia.utils.file.FileUtil
import timber.log.Timber
import java.net.SocketTimeoutException
import javax.inject.Inject
import kotlin.collections.ArrayList

/**
 * @author by alvinatin on 15/11/18.
 */
class UserIdentificationFormFinalFragment : BaseDaggerFragment(), UserIdentificationUploadImage.View, UserIdentificationFormActivity.Listener {
    private var loadingLayout: ConstraintLayout? = null
    private var mainLayout: ConstraintLayout? = null
    private var errorUploadLayout: RelativeLayout? = null
    private var resultImageKtp: ImageView? = null
    private var resultImageFace: ImageView? = null
    private var resultTextKtp: TextView? = null
    private var resultTextFace: TextView? = null
    private var bulletTextLayout: LinearLayout? = null
    private var info: TextView? = null
    private var subtitle: TextView? = null
    private var uploadButton: UnifyButton? = null
    private var stepperModel: UserIdentificationStepperModel? = null
    private var stepperListener: StepperListener? = null
    private var analytics: UserIdentificationCommonAnalytics? = null
    private var listRetake: ArrayList<Int> = arrayListOf()
    private var isSocketTimeoutException: Boolean = false

    private lateinit var remoteConfig: RemoteConfig

    private var retakeActionCode = NOT_RETAKE

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private val viewModelFragmentProvider by lazy { ViewModelProviders.of(this, viewModelFactory) }
    private val kycUploadViewModel by lazy { viewModelFragmentProvider.get(KycUploadViewModel::class.java) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (context is StepperListener) {
            stepperListener = context as StepperListener?
        }
        if (arguments != null && savedInstanceState == null) {
            stepperModel = arguments?.getParcelable(BaseStepperActivity.STEPPER_MODEL_EXTRA)
        } else if (savedInstanceState != null) {
            stepperModel = savedInstanceState.getParcelable(BaseUserIdentificationStepperFragment.EXTRA_KYC_STEPPER_MODEL)
        }
        if (activity != null) {
            analytics = UserIdentificationCommonAnalytics.createInstance(projectId)
        }

        remoteConfig = FirebaseRemoteConfigImpl(context)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putParcelable(BaseUserIdentificationStepperFragment.EXTRA_KYC_STEPPER_MODEL, stepperModel)
        super.onSaveInstanceState(outState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_user_identification_final, container, false)
        initView(view)
        encryptImage()
        setContentView()
        if (projectId == 4) //TradeIn project Id
            uploadButton?.setText(R.string.upload_button_tradein)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        analytics?.eventViewFinalForm()
        initObserver()
    }

    private fun initObserver() {
        kycUploadViewModel.kycResponseLiveData.observe(viewLifecycleOwner, androidx.lifecycle.Observer {
            when (it) {
                is Success -> {
                    sendSuccessTimberLog()
                    setKycUploadResultView(it.data)
                }
                is Fail -> {
                    hideLoading()
                    showUploadError()
                    setFailedResult(it.throwable)
                    sendErrorTimberLog(it.throwable)
                }
            }
        })

        kycUploadViewModel.encryptImageLiveData.observe(viewLifecycleOwner, androidx.lifecycle.Observer {
            when (it) {
                is Success -> {
                    uploadButton?.isEnabled = true
                    when (retakeActionCode) {
                        NOT_RETAKE -> {
                            //if liveness, upload the files immediately
                            if (!isKycSelfie) {
                                uploadKycFiles()
                            }
                        }
                        RETAKE_KTP -> {
                            goToLivenessOrSelfie()
                        }
                        RETAKE_FACE -> {
                            uploadKycFiles()
                        }
                    }
                }
                is Fail -> {
                    NetworkErrorHelper.showRedSnackbar(activity, resources.getString(R.string.error_text_image_fail_to_encrypt))
                }
            }
        })
    }

    private fun sendErrorTimberLog(throwable: Throwable) {
        if (!isKycSelfie) {
            ServerLogger.log(Priority.P2, "LIVENESS_UPLOAD_RESULT", mapOf("type" to "ErrorUpload",
                    "ktpPath" to stepperModel?.ktpFile.orEmpty(),
                    "facePath" to stepperModel?.faceFile.orEmpty(),
                    "tkpdProjectId" to projectId.toString(),
                    "stack_trace" to throwable.toString()
            ))
        } else {
            ServerLogger.log(Priority.P2, "SELFIE_UPLOAD_RESULT",
                    mapOf("type" to "ErrorUpload",
                            "ktpPath" to stepperModel?.ktpFile.orEmpty(),
                            "facePath" to stepperModel?.faceFile.orEmpty(),
                            "tkpdProjectId" to projectId.toString(),
                            "stack_trace" to throwable.toString()
                    ))
            analytics?.eventClickUploadPhotosTradeIn("failed")
        }
    }

    private fun sendSuccessTimberLog() {
        if (isKycSelfie) {
            ServerLogger.log(Priority.P2, "KYC_SELFIE_UPLOAD_RESULT",
                    mapOf("type" to "SuccessUpload",
                            "ktpPath" to stepperModel?.ktpFile.orEmpty(),
                            "facePath" to stepperModel?.faceFile.orEmpty(),
                            "tkpdProjectId" to projectId.toString()))
            analytics?.eventClickUploadPhotosTradeIn("success")
        }
    }

    override fun initInjector() {
        if (activity != null) {
            val daggerUserIdentificationComponent = DaggerUserIdentificationCommonComponent.builder()
                    .baseAppComponent((activity?.application as BaseMainApplication).baseAppComponent)
                    .build()
            daggerUserIdentificationComponent.inject(this)
        }
    }

    private fun openCameraView(viewMode: Int, requestCode: Int) {
        val intent = createIntent(context, viewMode)
        intent.putExtra(ApplinkConstInternalGlobal.PARAM_PROJECT_ID, projectId)
        startActivityForResult(intent, requestCode)
    }

    private fun openLivenessView() {
        val intent = RouteManager.getIntent(context, ApplinkConstInternalGlobal.LIVENESS_DETECTION)
        startActivityForResult(intent, KYCConstant.REQUEST_CODE_CAMERA_FACE)
    }

    private fun encryptImage() {
        if (isUsingEncrypt()) {
            uploadButton?.isEnabled = false
            kycUploadViewModel.encryptImageFace(stepperModel?.faceFile.toEmptyStringIfNull())
        }
    }

    private fun setContentView() {
        loadingLayout?.visibility = View.GONE
        if (isKycSelfie) {
            hideLoading()
            setKycSelfieView()
        } else {
            //if not using encryption, send immediately, else wait for encrypt and show loading
            if (!isUsingEncrypt()) {
                uploadKycFiles()
            } else {
                showLoading()
            }
        }
        if (activity is UserIdentificationFormActivity) {
            (activity as UserIdentificationFormActivity)
                    .updateToolbarTitle(getString(R.string.title_kyc_form_upload))
        }
    }

    private fun uploadKycFiles() {
        showLoading()
        stepperModel?.let {
            if (isSocketTimeoutException) {
                isSocketTimeoutException = false
            }
            kycUploadViewModel.uploadImages(it.ktpFile, it.faceFile, projectId.toString(), isUsingEncrypt())
        }
    }

    private fun setKycSelfieView() {
        if (activity is UserIdentificationFormActivity) {
            (activity as UserIdentificationFormActivity)
                    .updateToolbarTitle(getString(R.string.title_kyc_form_upload))
        }
        setResultViews(KycUrl.KTP_VERIF_OK, KycUrl.FACE_VERIF_OK, "", getString(R.string.form_final_info),
                ResourcesCompat.getColor(resources, com.tokopedia.unifyprinciples.R.color.Unify_N700_96, null),
                ResourcesCompat.getColor(resources, com.tokopedia.unifyprinciples.R.color.Unify_N700_96, null),
                getString(R.string.upload_button), null)
        generateLink()
        uploadButton?.setOnClickListener { v: View? ->
            analytics?.eventClickUploadPhotos()
            uploadKycFiles()
        }
    }

    private fun setKycUploadResultView(data: KycData) {
        if (data.isSuccessRegister) {
            deleteTmpFile(deleteKtp = true, deleteFace = true)
            activity?.setResult(Activity.RESULT_OK)
            stepperListener?.finishPage()
        } else {
            hideLoading()
            var imageKtp = KycUrl.KTP_VERIF_OK
            var imageFace = KycUrl.FACE_VERIF_OK
            var colorKtp: Int? = ResourcesCompat.getColor(resources, com.tokopedia.unifyprinciples.R.color.Unify_N700_96, null)
            var colorFace: Int? = ResourcesCompat.getColor(resources, com.tokopedia.unifyprinciples.R.color.Unify_N700_96, null)
            listRetake = data.listRetake
            if (!listRetake.isNullOrEmpty()) {
                for (i in data.listRetake.indices) {
                    when (data.listRetake[i]) {
                        KYCConstant.KTP_RETAKE -> {
                            imageKtp = KycUrl.KTP_VERIF_FAIL
                            colorKtp = null
                            setKtpRetakeButtonListener()
                        }
                        KYCConstant.FACE_RETAKE -> {
                            imageFace = KycUrl.FACE_VERIF_FAIL
                            colorFace = null
                            setFaceRetakeButtonListener()
                        }
                    }
                }
                if (listRetake.size == 2) {
                    setKtpFaceRetakeButtonListener()
                }
            }
            setResultViews(imageKtp, imageFace, data.app.title, data.app.subtitle, colorKtp, colorFace, data.app.button, data.listMessage)
            if (activity is UserIdentificationFormActivity) {
                (activity as UserIdentificationFormActivity)
                        .updateToolbarTitle(getString(R.string.title_kyc_form_fail_verification))
            }
        }
    }

    private fun setKtpRetakeButtonListener() {
        uploadButton?.setOnClickListener { v: View? ->
            analytics?.eventClickChangeKtpFinalFormPage()
            deleteTmpFile(deleteKtp = true, deleteFace = false)
            openCameraView(UserIdentificationCameraFragment.PARAM_VIEW_MODE_KTP, KYCConstant.REQUEST_CODE_CAMERA_KTP)
        }
    }

    private fun setFaceRetakeButtonListener() {
        uploadButton?.setOnClickListener { v: View? ->
            analytics?.eventClickChangeSelfieFinalFormPage()
            goToLivenessOrSelfie()
        }
    }

    private fun goToLivenessOrSelfie() {
        deleteTmpFile(deleteKtp = false, deleteFace = true)
        if (!isKycSelfie) {
            openLivenessView()
        } else {
            openCameraView(UserIdentificationCameraFragment.PARAM_VIEW_MODE_FACE, KYCConstant.REQUEST_CODE_CAMERA_FACE)
        }
    }

    private fun setKtpFaceRetakeButtonListener() {
        uploadButton?.setOnClickListener { v: View? ->
            analytics?.eventClickChangeKtpSelfieFinalFormPage()
            deleteTmpFile(deleteKtp = true, deleteFace = true)
            openCameraView(UserIdentificationCameraFragment.PARAM_VIEW_MODE_KTP, KYCConstant.REQUEST_CODE_CAMERA_KTP)
        }
    }

    private fun setResultViews(
            urlKtp: String,
            urlFace: String,
            subtitleText: String,
            infoText: String,
            colorKtp: Int?,
            colorFace: Int?,
            buttonText: String,
            listMessage: ArrayList<String>?
    ) {
        ImageHandler.LoadImage(resultImageKtp, urlKtp)
        ImageHandler.LoadImage(resultImageFace, urlFace)
        if (colorKtp != null) {
            resultTextKtp?.setTextColor(colorKtp)
        }
        if (colorFace != null) {
            resultTextFace?.setTextColor(colorFace)
        }
        subtitle?.gravity = Gravity.LEFT
        subtitle?.text = subtitleText
        info?.gravity = Gravity.LEFT
        info?.text = infoText
        uploadButton?.text = buttonText
        bulletTextLayout?.removeAllViewsInLayout()
        if (listMessage != null) {
            for (i in listMessage.indices) {
                bulletTextLayout?.let {
                    context?.let { context ->
                        (activity as UserIdentificationFormActivity).setTextViewWithBullet(listMessage[i], context, it)
                    }
                }
            }
        }
    }

    private val isKycSelfie: Boolean
        get() {
            try {
                if (UserIdentificationFormActivity.isSupportedLiveness) {
                    return remoteConfig.getBoolean(RemoteConfigKey.KYC_USING_SELFIE)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return true
        }

    private fun initView(view: View) {
        loadingLayout = view.findViewById(R.id.user_identification_final_loading_layout)
        mainLayout = view.findViewById(R.id.layout_main)
        resultImageKtp = view.findViewById(R.id.result_image_ktp)
        resultImageFace = view.findViewById(R.id.result_image_face)
        resultTextKtp = view.findViewById(R.id.result_text_ktp)
        resultTextFace = view.findViewById(R.id.result_text_face)
        bulletTextLayout = view.findViewById(R.id.layout_info_bullet)
        subtitle = view.findViewById(R.id.text_subtitle)
        info = view.findViewById(R.id.text_info)
        uploadButton = view.findViewById(R.id.upload_button)
        errorUploadLayout = view.findViewById(R.id.layout_kyc_upload_error)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK && data != null) {
            retakeAction(requestCode, data)
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun retakeAction(requestCode: Int, data: Intent) {
        when (requestCode) {
            KYCConstant.REQUEST_CODE_CAMERA_KTP -> {
                retakeActionCode = RETAKE_KTP
                stepperModel?.ktpFile = data.getStringExtra(KYCConstant.EXTRA_STRING_IMAGE_RESULT).toEmptyStringIfNull()
                if (isUsingEncrypt()) {
                    kycUploadViewModel.encryptImageKtp(stepperModel?.ktpFile.toEmptyStringIfNull())
                } else {
                    goToLivenessOrSelfie()
                }
            }
            KYCConstant.REQUEST_CODE_CAMERA_FACE -> {
                retakeActionCode = RETAKE_FACE
                if (!isKycSelfie) {
                    stepperModel?.faceFile = data.getStringExtra(ApplinkConstInternalGlobal.PARAM_FACE_PATH).toEmptyStringIfNull()
                } else {
                    stepperModel?.faceFile = data.getStringExtra(KYCConstant.EXTRA_STRING_IMAGE_RESULT).toEmptyStringIfNull()
                }
                if (isUsingEncrypt()) {
                    kycUploadViewModel.encryptImageFace(stepperModel?.faceFile.toEmptyStringIfNull())
                } else {
                    uploadKycFiles()
                }
            }
            else -> {
            }
        }
    }

    override fun getScreenName(): String = ""

    private fun generateLink() {
        val clickableSpan: ClickableSpan = object : ClickableSpan() {
            override fun onClick(widget: View) {
                analytics?.eventClickTermsFinalFormPage()
                RouteManager.route(activity, KycCommonUrl.APPLINK_TERMS_AND_CONDITION)
            }

            override fun updateDrawState(ds: TextPaint) {
                super.updateDrawState(ds)
                ds.isUnderlineText = false
                ds.color = resources.getColor(com.tokopedia.unifyprinciples.R.color.Unify_G400)
            }
        }
        val infoText = SpannableString(info?.text)
        val linked = resources.getString(R.string.terms_and_condition)
        val startIndex = info?.text.toString().indexOf(linked)
        infoText.setSpan(clickableSpan, startIndex, startIndex + linked.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        info?.highlightColor = Color.TRANSPARENT
        info?.movementMethod = LinkMovementMethod.getInstance()
        info?.setText(infoText, TextView.BufferType.SPANNABLE)
    }

    override val getContext: Context?
        get() = context

    override fun showLoading() {
        mainLayout?.visibility = View.GONE
        loadingLayout?.visibility = View.VISIBLE
        errorUploadLayout?.visibility = View.GONE
    }

    override fun hideLoading() {
        mainLayout?.visibility = View.VISIBLE
        loadingLayout?.visibility = View.GONE
        errorUploadLayout?.visibility = View.GONE
    }

    private fun showUploadError() {
        mainLayout?.visibility = View.GONE
        loadingLayout?.visibility = View.GONE
        errorUploadLayout?.visibility = View.VISIBLE
    }

    fun clickBackAction() {
        if (isSocketTimeoutException) {
            analytics?.eventClickBackConnectionTimeout()
        } else {
            if (!isKycSelfie) {
                if (listRetake.size == 1) {
                    when (listRetake[0]) {
                        KYCConstant.KTP_RETAKE -> {
                            run { analytics?.eventClickBackChangeKtpFinalFormPage() }
                            run { analytics?.eventClickBackChangeSelfieFinalFormPage() }
                        }
                        KYCConstant.FACE_RETAKE -> {
                            analytics?.eventClickBackChangeSelfieFinalFormPage()
                        }
                    }
                } else if (listRetake.size == 2) {
                    analytics?.eventClickBackChangeKtpSelfieFinalFormPage()
                }
            } else {
                analytics?.eventClickBackFinalForm()
            }
        }
    }

    override fun trackOnBackPressed() {}

    private fun setFailedResult(throwable: Throwable) {
        val errorCode = KycUploadErrorCodeUtil.getErrorCode(throwable)
        when (throwable) {
            is SocketTimeoutException -> {
                isSocketTimeoutException = true
                setViews(getString(R.string.kyc_upload_failed_reason_bad_network_title),
                        getString(R.string.kyc_upload_failed_reason_bad_network),
                        com.tokopedia.kyc_centralized.KycUrl.SCAN_FACE_FAIL_NETWORK)
            }
            else -> {
                setViews(getString(R.string.kyc_upload_failed_reason_general_title),
                        "${getString(R.string.kyc_upload_failed_reason_general)} ($errorCode)",
                        com.tokopedia.kyc_centralized.KycUrl.SCAN_FACE_FAIL_GENERAL)
            }
        }
        kyc_upload_error_button?.setOnClickListener {
            if (!isKycSelfie) {
                analytics?.eventClickConnectionTimeout()
                openLivenessView()
            } else {
                uploadKycFiles()
            }
        }
    }

    private fun setViews(failedReasonTitle: String, failedReason: String, failedImage: String) {
        kyc_upload_error_title?.text = failedReasonTitle
        kyc_upload_error_subtitle?.text = failedReason
        main_image?.let {
            ImageHandler.LoadImage(main_image, failedImage)
        }
    }

    fun deleteTmpFile(deleteKtp: Boolean, deleteFace: Boolean) {
        if (deleteKtp) FileUtil.deleteFile(stepperModel?.ktpFile)
        if (deleteFace) FileUtil.deleteFile(stepperModel?.faceFile)
    }

    private fun isUsingEncrypt(): Boolean {
        context?.let {
            return ImageEncryptionUtil.isUsingEncrypt(it)
        }
        return false
    }

    companion object {
        private var projectId = 0
        private const val NOT_RETAKE = 0
        private const val RETAKE_KTP = 1
        private const val RETAKE_FACE = 2

        fun createInstance(projectid: Int): Fragment {
            val fragment = UserIdentificationFormFinalFragment()
            projectId = projectid
            return fragment
        }
    }
}